target-gen/src/algorithm_binary.rs

```diff
@@ -128,12 +128,34 @@ impl AlgorithmBinary {
     }
 
     /// Assembles one huge binary blob as u8 values to write to RAM from the three sections.
-    pub(crate) fn blob(&self) -> Vec<u8> {
-        let mut blob = Vec::new();
+    pub(crate) fn blob(&self, arch: probe_rs_target::Architecture) -> Vec<u8> {
+        let mut blob = Vec::with_capacity(
+            (self.code_section.length + self.data_section.length + self.bss_section.length)
+                as usize,
+        );
+
+        // When flashing a target a specific blob will be prepended to this blob.
+        // This needs to be accounted for in the linking of the binary if it is not position independent
+        // Additionally, the sections might not be contigious due to memory limitations of the platform
+
+        let mut sections = [&self.code_section, &self.data_section, &self.bss_section];
+        sections.sort_by(|a, b| a.start.cmp(&b.start));
```

Comment:
I don't think that reordering these sections will work, right now the code in `probe-rs` assumes that these sections are in order.

What's the reason for this change? Do you have an example of a flash algorithm where the order was different?

Reply:
From my testing it works, although that testing has been limited to one or two different controllers.

As for the reasoning. I needed something like this to get some personal algorithm implementations to work correctly. They were needed for my job so unfortunately they can't be shared.

Reply:
I think we recently experienced issues with accidental ordering of regions. Do we have any other possibilities than ordering to make this work?

Reply:
Sorry, I haven't really been looking at this lately due to other priorities. What issues were you having?

Reply:
Basically this: https://github.com/probe-rs/probe-rs/pull/1745

Reply:
I don't really see how we can get around what caused me to sort them in the first place. Iirc probe-rs concatenates all the sections into a singular blob which it then writes to the specified address. To me this seems problematic as it assumes that the program is linked in the same order as probe-rs assumes it will be. If the program is position independent then all of a sudden the GOT needs to be updated with the addresses of the symbols it points to and the table itself may actually have to be moved depending on the linker script in use.

A possible way to solve this could be to change the target declaration schema slightly. Instead of representing the entire program as a single blob, define a mapping of addresses to blobs. If two or more sections are contiguous it is safe to concatenate them. Doing so also allows for better compatibility with the current schema.

Reply:
Are we speaking of how we map flash-algorithms to sections? Because this behavior where we concatenate sections should have been reverted in #1745.
As for flashing we definitely don't just concatenate data blocks. That would result in wildly wrong behavior. I am not sure I follow :/

Reply:
Maybe the function isn't in use any longer but iirc that is what `AlgorithmBinary::blob()‎` does (here)[]. Then when `FlashAlgorithm` reads the instructions from the config, it computes a value `addr_load` but only one is used in the end. Additionally, the instruction vector is never changed (with the exception of appending the architecture header).

Again, its been a while since I looked at all this but I think that is the reason why I solved it this way.